File: ui/src/main/java/ru/vertuos/ui/map/TiledMapParser.java
package ru.vertuos.ui.map;

import com.badlogic.gdx.maps.MapLayer;
import com.badlogic.gdx.maps.MapObject;
import com.badlogic.gdx.maps.MapObjects;
import com.badlogic.gdx.maps.MapProperties;
import com.badlogic.gdx.math.Rectangle;
import ru.vertuos.engine.map.object.CollisionableGameObject;
import ru.vertuos.engine.map.object.InclineObject;
import ru.vertuos.engine.map.object.PlatformObject;
import ru.vertuos.ui.contracts.DimensionUtils;

import java.util.ArrayList;
import java.util.List;


public class TiledMapParser {

    public List<CollisionableGameObject> parseCollisionObjects(
        MapLayer collisionLayer,
        float unitScale
    ) {
        MapObjects gdxMapObjects = collisionLayer.getObjects();

        List<CollisionableGameObject> collisionables = new ArrayList<>(gdxMapObjects.getCount());

        for (MapObject gdxMapObject : gdxMapObjects) {

            Rectangle bounds = extractBounds(gdxMapObject.getProperties());
            pixelsRectangleToMetresRectangle(bounds);
            scale(bounds, unitScale);

            switch (gdxMapObject.getName()) {
                case MapParsingContract.NAME_PLATFORM:
                    collisionables.add(new PlatformObject(bounds.x, bounds.y, bounds.width, bounds.height));
                    break;
                case MapParsingContract.NAME_INCLINE:
                    collisionables.add(new InclineObject(bounds.x, bounds.y, bounds.width, bounds.height));
                    break;
                case MapParsingContract.NAME_CIRCLED_INCLINE:
                    break;
                case MapParsingContract.NAME_HYPERBOLIC_INCLINE:
                    break;
                default:
                    break;
            }
        }
        return collisionables;
    }

    private Rectangle extractBounds(MapProperties properties) {
        float width = properties.get("width", Float.class);
        float height = properties.get("height", Float.class);
        float x = properties.get("x", Float.class);
        float y = properties.get("y", Float.class);
        return new Rectangle(width, height, x, y);
    }

    private void pixelsRectangleToMetresRectangle(Rectangle outRect) {
        outRect.x = DimensionUtils.pixelsToMetres(outRect.x);
        outRect.y = DimensionUtils.pixelsToMetres(outRect.y);
        outRect.width = DimensionUtils.pixelsToMetres(outRect.width);
        outRect.height = DimensionUtils.pixelsToMetres(outRect.height);
    }

    private void scale(Rectangle outRect, float scale) {
        outRect.x *= scale;
        outRect.y *= scale;
        outRect.width *= scale;
        outRect.height *= scale;
    }
}

File: ui/src/main/kotlin/ru/vertuos/ui/map/TiledMapParser.kt
package ru.vertuos.ui.map

import com.badlogic.gdx.maps.MapLayer
import com.badlogic.gdx.maps.MapProperties
import com.badlogic.gdx.math.Rectangle
import ru.vertuos.engine.map.CollisionableGameObject
import ru.vertuos.engine.map.inclines.CircledDownhillObject
import ru.vertuos.engine.map.inclines.CircledUphillObject
import ru.vertuos.engine.map.inclines.DownhillObject
import ru.vertuos.engine.map.inclines.UphillObject
import ru.vertuos.engine.map.obstacles.PlatformObject
import ru.vertuos.engine.math.Vector2f
import ru.vertuos.ui.contracts.pixelsToMetres
import java.util.ArrayList

class TiledMapParser {
    fun parseCollisionObjects(
        collisionLayer: MapLayer,
        unitScale: Float
    ): List<CollisionableGameObject> {

        val gdxMapObjects = collisionLayer.objects
        val collisionables = ArrayList<CollisionableGameObject>(gdxMapObjects.count)

        for (gdxMapObject in gdxMapObjects) {
            val bounds = extractBounds(gdxMapObject.properties)
            scale(bounds, unitScale)
            pixelsRectangleToMetresRectangle(bounds)
            bounds.apply {
                when (gdxMapObject.name) {
                    MapParsingContract.NAME_PLATFORM -> collisionables.add(
                        PlatformObject(Vector2f(x, y), Vector2f(width, height))
                    )
                    MapParsingContract.NAME_UPHILL -> collisionables.add(
                        UphillObject(Vector2f(x, y), Vector2f(width, height))
                    )
                    MapParsingContract.NAME_DOWNHILL -> collisionables.add(
                        DownhillObject(Vector2f(x, y), Vector2f(width, height))
                    )
                    MapParsingContract.NAME_CIRCLED_UPHILL -> collisionables.add(
                        CircledUphillObject(Vector2f(x, y), Vector2f(width, height))
                    )
                    MapParsingContract.NAME_CIRCLED_DOWNHILL -> collisionables.add(
                        CircledDownhillObject(Vector2f(x, y), Vector2f(width, height))
                    )
                    MapParsingContract.NAME_HYPERBOLIC_UPHILL -> collisionables.add(
                        UphillObject(Vector2f(x, y), Vector2f(width, height))
                    )
                    MapParsingContract.NAME_HYPERBOLIC_DOWNHILL -> collisionables.add(
                        DownhillObject(Vector2f(x, y), Vector2f(width, height))
                    )
                    else -> Unit
                }
            }
        }
        return collisionables
    }

    private fun extractBounds(properties: MapProperties): Rectangle = Rectangle(
        properties.x, properties.y, properties.width, properties.height
    )

    private fun pixelsRectangleToMetresRectangle(outRect: Rectangle) {
        outRect.x = outRect.x.pixelsToMetres()
        outRect.y = outRect.y.pixelsToMetres()
        outRect.width = outRect.width.pixelsToMetres()
        outRect.height = outRect.height.pixelsToMetres()
    }

    private fun scale(outRect: Rectangle, scale: Float) {
        outRect.x *= scale
        outRect.y *= scale
        outRect.width *= scale
        outRect.height *= scale
    }
}
